Dockerfile

```diff
@@ -39,65 +43,50 @@ RUN apt-get update \
     nodejs \
     npm \
  && apt-get clean \
- && rm -rf /var/lib/apt/lists/*
-
-RUN python3 -m pip install --upgrade setuptools pip build wheel
-RUN npm install --global yarn
-
-# copy everything except whats in .dockerignore, its a
-# compromise between needing to rebuild and maintaining
-# what needs to be part of the build
-COPY . /src/jupyterhub/
-WORKDIR /src/jupyterhub
+ && rm -rf /var/lib/apt/lists/* \
+ && python3 -m pip install --no-cache-dir --upgrade setuptools pip build wheel \
+ && npm install --global yarn \
+ # Build client component packages (they will be copied into ./share and
+ # packaged with the built wheel.)
+ && python3 -m build --wheel \
+ && python3 -m pip wheel --wheel-dir wheelhouse dist/*.whl
 
-# Build client component packages (they will be copied into ./share and
-# packaged with the built wheel.)
-RUN python3 -m build --wheel
-RUN python3 -m pip wheel --wheel-dir wheelhouse dist/*.whl
+FROM $BASE_IMAGE
+ENV DEBIAN_FRONTEND=noninteractive \
+    SHELL=/bin/bash \
+    LC_ALL=en_US.UTF-8 \
+    LANG=en_US.UTF-8 \
+    LANGUAGE=en_US.UTF-8
 
+EXPOSE 8000
 
-FROM $BASE_IMAGE
+LABEL maintainer="Jupyter Project <jupyter@googlegroups.com>"
+LABEL org.jupyter.service="jupyterhub"
 
-USER root
+WORKDIR /srv/jupyterhub
 
-ENV DEBIAN_FRONTEND=noninteractive
+COPY --from=builder /src/jupyterhub/wheelhouse /tmp/wheelhouse
 
 RUN apt-get update \
  && apt-get install -yq --no-install-recommends \
     ca-certificates \
     curl \
     gnupg \
     locales \
+    python-is-python3 \
     python3-pip \
     python3-pycurl \
     nodejs \
     npm \
+ && locale-gen $LC_ALL \
+ # always make sure pip is up to date!
+ && python3 -m pip install --no-cache-dir --upgrade setuptools pip \
+ && npm install -g configurable-http-proxy@^4.2.0 \
+ # install the wheels we built in the first stage
+ && python3 -m pip install --no-cache-dir /tmp/wheelhouse/* \
+ # clean cache and logs
  && apt-get clean \
- && rm -rf /var/lib/apt/lists/*
-
-ENV SHELL=/bin/bash \
-    LC_ALL=en_US.UTF-8 \
-    LANG=en_US.UTF-8 \
-    LANGUAGE=en_US.UTF-8
-
-RUN  locale-gen $LC_ALL
-
-# always make sure pip is up to date!
-RUN python3 -m pip install --no-cache --upgrade setuptools pip
-
-RUN npm install -g configurable-http-proxy@^4.2.0 \
- && rm -rf ~/.npm
-
-# install the wheels we built in the first stage
-COPY --from=builder /src/jupyterhub/wheelhouse /tmp/wheelhouse
-RUN python3 -m pip install --no-cache /tmp/wheelhouse/*
-
-RUN mkdir -p /srv/jupyterhub/
-WORKDIR /srv/jupyterhub/
-
-EXPOSE 8000
-
-LABEL maintainer="Jupyter Project <jupyter@googlegroups.com>"
-LABEL org.jupyter.service="jupyterhub"
+ && rm -rf /var/lib/apt/lists/* /var/log/* /var/tmp/* ~/.npm \
+ && find / -type d -name '*__pycache__' -prune -exec rm -rf {} \;
```

Comment:
This is probably simplified with

```
ENV PYTHONDONTWRITEBYTECODE=1
```

prior to installation, without any pruning. This also has the benefit of preventing re-writing these files on every container startup.

Removing bytecode generally has a (small) negative impact on startup performance, which I guess is traded for pull-time performance - so the first startup+pull _might_ be quicker, but all startups after that are slower. This impact is doubled if PYTHONDONTWRITEBYTECODE is not set in the container, because all the missing bytecode files will be written out - never to be used - on _every start_.